File: src/lib/buildInfo.ts
/**
 * Build info IPC wrapper + cache.
 *
 * `version` / `buildNumber` / `buildTime` are baked at compile time by
 * `src-tauri/build.rs` and exposed via the `get_build_info` Tauri command.
 * The values never change for a running process, so we cache the first
 * fetch and serve subsequent calls from memory.
 */
import { invoke } from './ipc';

export interface BuildInfo {
	version: string;
	buildNumber: string;
	buildTime: string;
	/** Single source of truth for gating, resolved in Rust. */
	channel: 'dev' | 'release';
	/**
	 * Resolved flag booleans. On `release` the backend omits `kind=dev`
	 * entries (feature-flags.md §3.5); never carries registry metadata.
	 */
	flags: Record<string, boolean>;
}

let cached: BuildInfo | null = null;
let inflight: Promise<BuildInfo> | null = null;

export async function getBuildInfo(): Promise<BuildInfo> {
	if (cached) return cached;
	if (inflight) return inflight;
	inflight = invoke<BuildInfo>('get_build_info').then((info) => {
		cached = info;
		inflight = null;
		return info;
	});
	return inflight;
}

File: src-tauri/src/commands/build_info.rs
//! `get_build_info` command — exposes compile-time build metadata to the
//! frontend (version, build number, build time).
//!
//! Build number / time are baked at compile time by `build.rs` via the
//! `VELLIS_BUILD_NUMBER` and `VELLIS_BUILD_TIME` env vars, so calling this
//! command is a constant-time lookup. `channel` / `flags` are resolved in
//! Rust (feature-flags.md §3.5) — the frontend never re-evaluates them.

use std::collections::BTreeMap;

use serde::Serialize;

use crate::features;

#[derive(Serialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BuildInfo {
    pub version: &'static str,
    pub build_number: &'static str,
    pub build_time: &'static str,
    /// `"release"` | `"dev"` — single source of truth for the frontend.
    pub channel: &'static str,
    /// Resolved flag booleans. On `release`, `kind=dev` entries are
    /// omitted (§3.5 payload minimization); registry metadata
    /// (owner/removal/…) never crosses IPC.
    pub flags: BTreeMap<&'static str, bool>,
}

#[tauri::command]
pub fn get_build_info() -> BuildInfo {
    BuildInfo {
        version: env!("CARGO_PKG_VERSION"),
        build_number: env!("VELLIS_BUILD_NUMBER"),
        build_time: env!("VELLIS_BUILD_TIME"),
        channel: features::current_channel().as_str(),
        flags: features::flags_map(),
    }
}
